lib/AST/UnqualifiedLookup.cpp

```diff
@@ -553,6 +557,49 @@ bool ASTScopeDeclConsumerForUnqualifiedLookup::consume(
   for (auto *value: values) {
     if (factory.isOriginallyTypeLookup && !isa<TypeDecl>(value))
       continue;
+
+    // Try to resolve the base for unqualified instance member
+    // references. This is used by lookInMembers().
+    if (auto *var = dyn_cast<VarDecl>(value)) {
+      // We have a binding named 'self'.
+      //
+      // There are three possibilities:
+      //
+      // 1) This binding is the 'self' parameter of a method,
+      // 2) This binding is a bona-fide 'self' capture, meaning a capture
+      //    list entry named 'self' with initial value expression 'self',
+      // 3) None of the above.
+      if (var->getName() == factory.Ctx.Id_self) {
+        // How we handle these cases depends on whether we've already seen
+        // another 'self' binding.
+        if (candidateSelfDC == nullptr) {
+          // We haven't seen one yet, so record it.
+          if (var->isSelfParameter())
+            candidateSelfDC = var->getDeclContext();
+          else if (var->isSelfParamCapture())
+            candidateSelfDC = var->getParentCaptureList()->getClosureBody();
+        } else {
+          // If we see a binding named 'self' that is not a bona-fide
+          // 'self', we have to forget about the previous 'self' capture
+          // because it's not going to be the right one for accessing
+          // instance members of the innermost nominal type. Eg,
+          //
+          // class C {
+          //   func bar() {}
+          //   func foo() {
+          //     _ { [self=12] { [self] bar() } }
+          //   }
+          // }
+          //
+          // Instead, we're going to move on and look for the next-innermost
+          // 'self' binding.
+          if (!var->isSelfParameter() &&
+              !var->isSelfParamCapture())
+            candidateSelfDC = nullptr;
+        }
+      }
+    }
+
```

Comment:
The comments are great! But, at this point, this function is too long for me to process, even though most of it is comments. I suspect you could break it up into very manageable pieces. Maybe split off the body of the for loop into one piece, and split the two halves of the `if` on 563? That's just what I would try. You can find something better, I bet. Something that communicates what's really going on here, succinctly at the upper levels.

Reply:
I split off the body of the `if` statement into a new `maybeUpdateSelfDC()` method.